### PR TITLE
 B-20229 refactor fuel pricer

### PR DESCRIPTION
**Pull request by deandreJones**
[B-20229](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A964589)

# Summary

Update FRA weekly effective period
 # Background
the app pulls the diesel prices every day looking for newly published prices
EIA publishes new prices generally every Monday, before this change they would just be effective until the next publication.
This change makes it so that when new diesel prices are published they are only valid from tueday( or the day after they are published0 through the following Monday 

# to test
1. run `make tasks_save_ghc_fuel_price_data` to get the latest pricing data for your test
2. create a new move
3. approve move
4. access prime and create a payment request for fuel surcharge 
5. login as TIO - review the payment request
6. validate that it selected the correct price based on the actual pickup date being between the effective and End dates in the DB GHC_DIESEL_FUEL_PRICES

- may want to manually enter a few extra rows verify it works and making the actual pickup date land on the effecting and end dates 
Screenshot 2024-06-07 at 2 26 11 PM
Screenshot 2024-06-07 at 2 57 05 PM

image

### CODE REVIEW

File: scripts/run-server-test [hunk 1]
```diff
@@ -47,6 +47,7 @@ else
   gotest_args+=("-parallel" "8")
   gotest_args+=("-failfast")
 fi
+#gotest_args+=("-ldflags=-extldflags=-Wl,-ld_classic")
```
Comment (cameroncaci) [inline scripts/run-server-test:51]: Could we comment why this is commented?
https://caci-bits.slack.com/archives/C05DX3AG6EL/p1717687099941339

**Review by cameroncaci — CHANGES_REQUESTED** [review]

**Review by cameroncaci — COMMENTED** [review]
After creating a new move as a customer, logging into officelocal as HAFC GBLOC multirole, selecting to become a service counselor, and then attempting to view and edit orders, I'm encountering errors.

This was encountered on the Chrome browser

image


image
image

File: scripts/run-server-test (hunk 1 above)
Comment (deandreJones) [reply scripts/run-server-test:51]: https://github.com/golang/go/issues/61229#issuecomment-1988965927

ill add more to the line- but in short- 
our (Mac) build is broken for the go linker on the go version we are using- this forces the use of the older linker

**Review by danieljordan-caci — CHANGES_REQUESTED** [review]

File: migrations/app/schema/20240604203456_add_effective_expiraton_date_ghc_fuel.up.sql [hunk 2]
```diff
@@ -0,0 +1,8 @@
+ALTER Table ghc_diesel_fuel_prices
+ADD column IF NOT EXISTS effective_date date,
+ADD column IF NOT EXISTS end_date date;
```
[the remark this answers is not in the document]
Comment (deandreJones) [reply migrations/app/schema/20240604203456_add_effective_expiraton_date_ghc_fuel.up.sql:3]: low risk on this transaction- all current rows must have publication date - and only added weekly - so not a lot of records are being updated on this

[the remark this answers is not in the document]
Comment (danieljordan-caci) [reply migrations/app/schema/20240604203456_add_effective_expiraton_date_ghc_fuel.up.sql:3]: Gotcha. I'll resolve this convo then! Just figured I'd ask!

File: pkg/payment_request/service_param_value_lookups/eia_fuel_price_lookup.go [hunk 3]
```diff
@@ -25,13 +25,22 @@ func (r EIAFuelPriceLookup) lookup(appCtx appcontext.AppContext, _ *ServiceItemP
 		return "", fmt.Errorf("could not find actual pickup date for MTOShipment [%s]", r.MTOShipment.ID)
 	}
 
-	// Find the GHCDieselFuelPrice object with the closest prior PublicationDate to the ActualPickupDate of the MTOShipment in question
+	// Find the GHCDieselFuelPrice object effective before the shipment's ActualPickupDate and ends after the ActualPickupDate
 	var ghcDieselFuelPrice models.GHCDieselFuelPrice
-	err := db.Where("publication_date <= ?", actualPickupDate).Order("publication_date DESC").Last(&ghcDieselFuelPrice)
+	err := db.Where("? BETWEEN effective_date and end_date", actualPickupDate).Order("publication_Date DESC").First(&ghcDieselFuelPrice) //only want the first published price per week
```
[the remark this answers is not in the document]
Comment (danieljordan-caci) [reply pkg/payment_request/service_param_value_lookups/eia_fuel_price_lookup.go:30]: Such a sweetie

**Comment by TevinAdams** [comment]
I had a pickup date 2 days after the latest ghc_diesel_fuel_prices. So my pick-up date was 06-12-2024 and in the DB the latest record had an effective date of 2024-06-04 and an end date of 2024-06-10. This records fuel_price_in_milicents was selected even though the pick up date was after its end date. It looks like in this case it picked the latest. Is that expected? Also how do we get these updated prices? If June 12th is 2 days from now but there is no ghc_diesel_fuel_prices with that date in its range what is that a likely situation and what do we expect to happen in that case?

**Comment by deandreJones** [comment]
> I had a pickup date 2 days after the latest ghc_diesel_fuel_prices. So my pick-up date was 06-12-2024 and in the DB the latest record had an effective date of 2024-06-04 and an end date of 2024-06-10. This records fuel_price_in_milicents was selected even though the pick up date was after its end date. It looks like in this case it picked the latest. Is that expected? Also how do we get these updated prices? If June 12th is 2 days from now but there is no ghc_diesel_fuel_prices with that date in its range what is that a likely situation and what do we expect to happen in that case?

Yes- this is what it should do-
if there isn't a newly published rate- then it should grab the last published

**Comment by TevinAdams** [comment]
I changed the pickup date to 06/02/2024 and it chose the ghc fuel price for date range 2020-07-21 - 2020-07-27. Not sure If thats expected.
image

**Review by danieljordan-caci — APPROVED** [review]
Tested locally and seems good to go. You have some failing tests, but I'd bet they're flaky.

**Review by traskowskycaci — APPROVED** [review]
Replicated your results locally, looks good!

**Pull request merged**